subsys/net/ip/tcp.c

```diff
@@ -2428,6 +2437,11 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 
 	k_mutex_lock(&conn->lock, K_FOREVER);
 
+	/* Connection context was already freed. */
+	if (conn->state == TCP_UNUSED) {
+		return NET_DROP;
```

Comment:
nit: you may want to add to the comment that a k_mutex_unlock() is not needed before return NET_DROP. (I'm assuming it's not)

Also, we are relying that nothing else touches this memory after k_mem_slab_free(). Can anything call tcp_conn_alloc() and use this memory in some other race condition?

Reply:
> 
> In any case however, do you have any idea how could we protect against such case? I can't really think of any good countermeasure other than a complete redesign on how connection contexts are allocated (i. e. connection-related mutexes should only be initialized once, on boot).

I can't see any simple ways either.